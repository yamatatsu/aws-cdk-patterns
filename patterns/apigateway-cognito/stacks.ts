import {
  App,
  Stack,
  StackProps,
  aws_lambda as lambda,
  aws_apigateway as apigateway,
  aws_cognito as cognito,
} from "aws-cdk-lib"

export class ApigatewayCognito extends Stack {
  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props)

    const userPool = new cognito.UserPool(this, "UserPool", {
      selfSignUpEnabled: true,
      /**
       * emailのみにすると SignUp API にて UsernameExistsException によるユーザー存在確認ができてしまう。
       * `username: true` を追加することで メール疎通以前においてメアド重複エラーが発生しなくなるため、ユーザー存在確認ができなくなる。
       * ただし username の入力が必須になる。変更はできないので変更機能の実装は不要。
       * 変更もしたい場合は preferred_username を使う。
       *
       * SignUp API の呼び出しサンプル
       *
       * curl https://cognito-idp.ap-northeast-1.amazonaws.com/ \
       *   -H 'Content-Type: application/x-amz-json-1.1' \
       *   -H 'x-amz-target: AWSCognitoIdentityProviderService.SignUp' \
       *   -d '{"ClientId":"xxxxxxxxxxxxxxxxx","Username":"hogehoge","UserAttributes": [{"Name":"email","Value":"xxxxxxxxxx@gmail.com"}],"Password": "xxxxxxxxxx","ValidationData":null}'
       *
       * curl https://cognito-idp.ap-northeast-1.amazonaws.com/ \
       *   -H 'Content-Type: application/x-amz-json-1.1' \
       *   -H 'x-amz-target: AWSCognitoIdentityProviderService.ConfirmSignUp' \
       *   -d '{"ClientId":"xxxxxxxxxxxxxxxxx","Username":"hogehoge","ConfirmationCode":"xxxxxx"}'
       */
      signInAliases: { username: true, email: true },
      userVerification: {
        emailStyle: cognito.VerificationEmailStyle.CODE,
      },
    })
    const userPoolClient = new cognito.UserPoolClient(this, "UserPoolClient", {
      userPool,
      authFlows: { userSrp: true },
    })

    new cognito.UserPoolDomain(this, "UserPoolDomain", {
      userPool,
      cognitoDomain: { domainPrefix: `yamatatsu-apigateway-example` },
    })

    const handler = new lambda.Function(this, "Lambda", {
      code: new lambda.AssetCode("./lambda"),
      handler: "index.handler",
      runtime: lambda.Runtime.NODEJS_14_X,
    })

    const restApi = new apigateway.RestApi(this, "RestApi", {
      restApiName: "ApigatewayCognito_RestApi",
      deploy: true,
    })

    const cognitoAuthrizer = new apigateway.CfnAuthorizer(
      this,
      "CfnAuthorizer",
      {
        name: "Cognito-Userpool",
        type: "COGNITO_USER_POOLS",
        restApiId: restApi.restApiId,
        identitySource: "method.request.header.x-ytoken",
        providerArns: [userPool.userPoolArn],
      },
    )

    restApi.root.addProxy({
      defaultIntegration: new apigateway.LambdaIntegration(handler),
      defaultMethodOptions: {
        authorizer: {
          authorizationType: apigateway.AuthorizationType.COGNITO,
          authorizerId: cognitoAuthrizer.ref,
        },
      },
    })
  }
}
