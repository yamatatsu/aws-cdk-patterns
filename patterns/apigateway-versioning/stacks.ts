import * as cdk from "@aws-cdk/core"
import * as lambda from "@aws-cdk/aws-lambda"
import * as apigateway from "@aws-cdk/aws-apigateway"

export class ApigatewayVersioning extends cdk.Stack {
  constructor(parent: cdk.App, id: string, props?: cdk.StackProps) {
    super(parent, id, props)

    const handler = new lambda.Function(this, "Lambda", {
      code: new lambda.AssetCode("./lambda"),
      handler: "index.handler",
      runtime: lambda.Runtime.NODEJS_12_X,
      tracing: lambda.Tracing.ACTIVE,
      currentVersionOptions: {
        removalPolicy: cdk.RemovalPolicy.RETAIN,
      },
    })

    const restApi = new apigateway.RestApi(this, "RestApi", {
      restApiName: "ApiGatewayVersioning",
      endpointTypes: [apigateway.EndpointType.REGIONAL],
      deployOptions: {
        loggingLevel: apigateway.MethodLoggingLevel.INFO,
      },
      defaultIntegration: new apigateway.LambdaIntegration(handler),
    })

    const fixVersion = genFixVersion(this, handler, restApi)
    fixVersion("v1.0.0", "6")
    fixVersion("v1.0.1", "7")
    fixVersion("v1.0.2", handler.currentVersion.version)
  }
}

const genFixVersion = (
  scope: cdk.Stack,
  handler: lambda.IFunction,
  restApi: apigateway.IRestApi,
) => (versionName: string, version: string) => {
  const _versionName = versionName.replace(/\./g, "_")

  const alias = lambda.Version.fromVersionAttributes(
    scope,
    `lambdaVersion_${_versionName}`,
    { lambda: handler, version },
  ).addAlias(_versionName)

  restApi.root.addResource(versionName).addProxy({
    defaultIntegration: new apigateway.LambdaIntegration(alias),
  })
}
