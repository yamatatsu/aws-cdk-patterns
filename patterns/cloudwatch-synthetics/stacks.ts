// FIXME: RCにはまだ synthetics が実装されていない。実装を待つ。
// import {
//   App,
//   Stack,
//   StackProps,
//   RemovalPolicy,
//   Duration,
//   aws_synthetics as synthetics,
//   aws_s3 as s3,
// } from "aws-cdk-lib"

// type Props = StackProps & {
//   code: synthetics.Code
// }

// export class CloudwatchSynthetics extends Stack {
//   constructor(parent: App, id: string, props: Props) {
//     super(parent, id, props)

//     const bucket = new s3.Bucket(this, "Bucket", {
//       removalPolicy: RemovalPolicy.DESTROY,
//     })

//     new synthetics.Canary(this, "Canary", {
//       artifactsBucketLocation: { bucket },
//       // role?: iam.IRole
//       timeToLive: Duration.minutes(5),
//       schedule: synthetics.Schedule.once(),
//       // startAfterCreation?: boolean
//       // successRetentionPeriod: Duration.days(7),
//       // failureRetentionPeriod: Duration.days(0),
//       // canaryName?: string
//       runtime: synthetics.Runtime.SYNTHETICS_NODEJS_PUPPETEER_3_0,
//       test: synthetics.Test.custom({
//         code: props.code,
//         handler: "index.handler",
//       }),
//     })
//   }
// }
