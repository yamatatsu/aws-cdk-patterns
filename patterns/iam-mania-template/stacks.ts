import { App, Arn, aws_iam, Stack, StackProps } from "aws-cdk-lib"

export class IamManiaAdmin extends Stack {
  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props)

    const ipRestrictionAndMFAForcePolicy = new aws_iam.ManagedPolicy(
      this,
      "IpRestrictionAndMFAForcePolicy",
      {
        managedPolicyName: "IpRestrictionAndMFAForcePolicy",
        statements: [
          // [DENY]IP制限
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.DENY,
            actions: ["*"],
            resources: ["*"],
            // 任意のIPに変更すること
            conditions: { NotIpAddress: { "aws:SourceIp": ["8.8.8.8/32"] } },
          }),
          // [DENY]MFA必須化
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.DENY,
            notActions: ["iam:*"],
            resources: ["*"],
            conditions: {
              BoolIfExists: { "aws:MultiFactorAuthPresent": "false" },
            },
          }),
          // [ALLOW]全ユーザーのリスト表示
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.ALLOW,
            actions: ["iam:ListUsers", "iam:ListVirtualMFADevices"],
            resources: ["*"],
          }),
        ],
      },
    )
    const changeOwnPasswordPolicy = new aws_iam.ManagedPolicy(
      this,
      "ChangeOwnPasswordPolicy",
      {
        managedPolicyName: "ChangeOwnPasswordPolicy",
        statements: [
          // [ALLOW]自身のパスワード変更＆MFA有効化ポリシー
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.ALLOW,
            actions: [
              "iam:ChangePassword",
              "iam:CreateAccessKey",
              "iam:CreateVirtualMFADevice",
              "iam:DeactivateMFADevice",
              "iam:DeleteAccessKey",
              "iam:DeleteVirtualMFADevice",
              "iam:EnableMFADevice",
              "iam:GetAccountPasswordPolicy",
              "iam:UpdateAccessKey",
              "iam:UpdateSigningCertificate",
              "iam:UploadSigningCertificate",
              "iam:UpdateLoginProfile",
              "iam:ResyncMFADevice",
            ],
            resources: [
              Arn.format(
                {
                  service: "iam",
                  region: "",
                  resource: "user",
                  resourceName: "${aws:username}",
                },
                this,
              ),
              Arn.format(
                {
                  service: "iam",
                  region: "",
                  resource: "mfa",
                  resourceName: "${aws:username}",
                },
                this,
              ),
            ],
          }),
          // [ALLOW]全ユーザーのリスト表示
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.ALLOW,
            actions: ["iam:ListUsers", "iam:ListVirtualMFADevices"],
            resources: ["*"],
          }),
        ],
      },
    )
    const denyNetworkAccessPolicy = new aws_iam.ManagedPolicy(
      this,
      "DenyNetworkAccessPolicy",
      {
        managedPolicyName: "DenyNetworkAccessPolicy",
        statements: [
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.DENY,
            actions: [
              "ec2:AcceptVpcEndpointConnections",
              "ec2:AssociateDhcpOptions",
              "ec2:AssociateRouteTable",
              "ec2:AssociateSubnetCidrBlock",
              "ec2:AssociateVpcCidrBlock",
              "ec2:AttachInternetGateway",
              "ec2:AttachNetworkInterface",
              "ec2:AttachVpnGateway",
              "ec2:CreateCustomerGateway",
              "ec2:CreateDefaultSubnet",
              "ec2:CreateDefaultVpc",
              "ec2:CreateDhcpOptions",
              "ec2:CreateEgressOnlyInternetGateway",
              "ec2:CreateInternetGateway",
              "ec2:CreateNetworkAcl",
              "ec2:CreateNetworkAclEntry",
              "ec2:CreateRoute",
              "ec2:CreateRouteTable",
              "ec2:CreateSecurityGroup",
              "ec2:CreateSubnet",
              "ec2:CreateVpc",
              "ec2:CreateVpcEndpoint",
              "ec2:CreateVpcEndpointConnectionNotification",
              "ec2:CreateVpcEndpointServiceConfiguration",
              "ec2:CreateVpnConnection",
              "ec2:CreateVpnConnectionRoute",
              "ec2:CreateVpnGateway",
              "ec2:DeleteEgressOnlyInternetGateway",
              "ec2:DeleteNatGateway",
              "ec2:DeleteNetworkInterface",
              "ec2:DeleteNetworkInterfacePermission",
              "ec2:DeletePlacementGroup",
              "ec2:DeleteSubnet",
              "ec2:DeleteVpc",
              "ec2:DeleteVpcEndpointConnectionNotifications",
              "ec2:DeleteVpcEndpoints",
              "ec2:DeleteVpcEndpointServiceConfigurations",
              "ec2:DeleteVpnConnection",
              "ec2:DeleteVpnConnectionRoute",
              "ec2:DeleteVpnGateway",
              "ec2:DetachInternetGateway",
              "ec2:DetachNetworkInterface",
              "ec2:DetachVpnGateway",
              "ec2:DisableVgwRoutePropagation",
              "ec2:DisableVpcClassicLinkDnsSupport",
              "ec2:DisassociateRouteTable",
              "ec2:DisassociateSubnetCidrBlock",
              "ec2:DisassociateVpcCidrBlock",
              "ec2:EnableVgwRoutePropagation",
              "ec2:EnableVpcClassicLinkDnsSupport",
              "ec2:ModifyNetworkInterfaceAttribute",
              "ec2:ModifySubnetAttribute",
              "ec2:ModifyVpcAttribute",
              "ec2:ModifyVpcEndpoint",
              "ec2:ModifyVpcEndpointConnectionNotification",
              "ec2:ModifyVpcEndpointServiceConfiguration",
              "ec2:ModifyVpcEndpointServicePermissions",
              "ec2:ModifyVpcPeeringConnectionOptions",
              "ec2:ModifyVpcTenancy",
              "ec2:MoveAddressToVpc",
              "ec2:RejectVpcEndpointConnections",
              "ec2:ReplaceNetworkAclAssociation",
              "ec2:ReplaceNetworkAclEntry",
              "ec2:ReplaceRoute",
              "ec2:ReplaceRouteTableAssociation",
              "ec2:ResetNetworkInterfaceAttribute",
              "ec2:RestoreAddressToClassic",
              "ec2:UpdateSecurityGroupRuleDescriptionsEgress",
              "ec2:UpdateSecurityGroupRuleDescriptionsIngress",
              "directconnect:*",
            ],
            resources: ["*"],
          }),
          new aws_iam.PolicyStatement({
            effect: aws_iam.Effect.DENY,
            actions: [
              "ec2:AcceptVpcPeeringConnection",
              "ec2:AttachClassicLinkVpc",
              "ec2:AuthorizeSecurityGroupEgress",
              "ec2:AuthorizeSecurityGroupIngress",
              "ec2:CreateVpcPeeringConnection",
              "ec2:DeleteCustomerGateway",
              "ec2:DeleteDhcpOptions",
              "ec2:DeleteInternetGateway",
              "ec2:DeleteNetworkAcl",
              "ec2:DeleteNetworkAclEntry",
              "ec2:DeleteRoute",
              "ec2:DeleteRouteTable",
              "ec2:DeleteSecurityGroup",
              "ec2:DeleteVolume",
              "ec2:DeleteVpcPeeringConnection",
              "ec2:DetachClassicLinkVpc",
              "ec2:DisableVpcClassicLink",
              "ec2:EnableVpcClassicLink",
              "ec2:GetConsoleScreenshot",
              "ec2:RejectVpcPeeringConnection",
              "ec2:RevokeSecurityGroupEgress",
              "ec2:RevokeSecurityGroupIngress",
              "ec2:AcceptTransitGatewayVpcAttachment",
              "ec2:AssociateTransitGatewayRouteTable",
              "ec2:CreateTransitGateway",
              "ec2:CreateTransitGatewayRoute",
              "ec2:CreateTransitGatewayRouteTable",
              "ec2:CreateTransitGatewayVpcAttachment",
              "ec2:DeleteTransitGateway",
              "ec2:DeleteTransitGatewayRoute",
              "ec2:DeleteTransitGatewayRouteTable",
              "ec2:DeleteTransitGatewayVpcAttachment",
              "ec2:DescribeTransitGatewayAttachments",
              "ec2:DescribeTransitGatewayRouteTables",
              "ec2:DescribeTransitGatewayVpcAttachments",
              "ec2:DescribeTransitGateways",
              "ec2:DisableTransitGatewayRouteTablePropagation",
              "ec2:DisassociateTransitGatewayRouteTable",
              "ec2:EnableTransitGatewayRouteTablePropagation",
              "ec2:ExportTransitGatewayRoutes",
              "ec2:GetTransitGatewayAttachmentPropagations",
              "ec2:GetTransitGatewayRouteTableAssociations",
              "ec2:GetTransitGatewayRouteTablePropagations",
              "ec2:ModifyTransitGatewayVpcAttachment",
              "ec2:RejectTransitGatewayVpcAttachment",
              "ec2:ReplaceTransitGatewayRoute",
              "ec2:SearchTransitGatewayRoutes",
            ],
            resources: ["*"],
          }),
        ],
      },
    )

    const adminGroup = new aws_iam.Group(this, "AdminGroup", {
      groupName: "Admins",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess"),
        ipRestrictionAndMFAForcePolicy,
      ],
    })
    const networkAdminGroup = new aws_iam.Group(this, "NetworkAdminGroup", {
      groupName: "NetworkAdmins",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName(
          "job-function/NetworkAdministrator",
        ),
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("ReadOnlyAccess"),
        ipRestrictionAndMFAForcePolicy,
        changeOwnPasswordPolicy,
      ],
    })
    const developerGroup = new aws_iam.Group(this, "DeveloperGroup", {
      groupName: "Developers",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("PowerUserAccess"),
        ipRestrictionAndMFAForcePolicy,
        changeOwnPasswordPolicy,
        denyNetworkAccessPolicy,
      ],
    })
    const operatorGroup = new aws_iam.Group(this, "OperatorGroup", {
      groupName: "Operators",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("AWSSupportAccess"),
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("ReadOnlyAccess"),
        ipRestrictionAndMFAForcePolicy,
        changeOwnPasswordPolicy,
      ],
    })
    const billingUserGroup = new aws_iam.Group(this, "BillingUserGroup", {
      groupName: "BillingUsers",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("job-function/Billing"),
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("ReadOnlyAccess"),
        ipRestrictionAndMFAForcePolicy,
        changeOwnPasswordPolicy,
      ],
    })

    const user = new aws_iam.User(this, "user", {
      userName: "hoge",
      groups: [adminGroup],
    })

    new aws_iam.Role(this, "AdminRole", {
      roleName: "AdminRole",
      assumedBy: new aws_iam.ArnPrincipal(user.userArn),
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess"),
      ],
    })
  }
}
