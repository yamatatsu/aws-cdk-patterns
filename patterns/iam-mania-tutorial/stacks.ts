import { App, Arn, aws_iam, Stack, StackProps } from "aws-cdk-lib"

export class IamManiaTutorial extends Stack {
  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props)

    const managedPolicy = new aws_iam.ManagedPolicy(this, "ManagedPolicy", {
      managedPolicyName: "IPAddressRestriction",
      statements: [
        new aws_iam.PolicyStatement({
          effect: aws_iam.Effect.DENY,
          actions: ["*"],
          resources: ["*"],
          conditions: { NotIpAddress: { "aws:SourceIp": ["8.8.8.8/32"] } },
        }),
      ],
    })

    const group = new aws_iam.Group(this, "Group", {
      groupName: "ViewOnlyUsers",
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("ReadOnlyAccess"),
        managedPolicy,
      ],
    })

    const user = new aws_iam.User(this, "User", {
      userName: "test-user01",
      groups: [group],
    })

    new aws_iam.Role(this, "Role", {
      roleName: "SwitchRoleForTest01",
      assumedBy: new aws_iam.ArnPrincipal(user.userArn),
      managedPolicies: [
        aws_iam.ManagedPolicy.fromAwsManagedPolicyName("AdministratorAccess"),
      ],
    })
  }
}
