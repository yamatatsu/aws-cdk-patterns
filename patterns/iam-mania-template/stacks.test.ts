import { SynthUtils } from "@aws-cdk/assert"
import { App } from "aws-cdk-lib"
import { IamManiaAdmin } from "./stacks"

test("snapshot test", () => {
  const app = new App()

  const target = new IamManiaAdmin(app, "Target")

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot()
})
