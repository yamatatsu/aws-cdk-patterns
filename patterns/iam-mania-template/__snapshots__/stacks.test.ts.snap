// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Resources": Object {
    "AdminGroupCD52047A": Object {
      "Properties": Object {
        "GroupName": "Admins",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
          Object {
            "Ref": "IpRestrictionAndMFAForcePolicyA06EC28F",
          },
        ],
      },
      "Type": "AWS::IAM::Group",
    },
    "AdminRole38563C57": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "user2C2B57AE",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
        "RoleName": "AdminRole",
      },
      "Type": "AWS::IAM::Role",
    },
    "BillingUserGroup901A75C7": Object {
      "Properties": Object {
        "GroupName": "BillingUsers",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/job-function/Billing",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/ReadOnlyAccess",
              ],
            ],
          },
          Object {
            "Ref": "IpRestrictionAndMFAForcePolicyA06EC28F",
          },
          Object {
            "Ref": "ChangeOwnPasswordPolicy67755262",
          },
        ],
      },
      "Type": "AWS::IAM::Group",
    },
    "ChangeOwnPasswordPolicy67755262": Object {
      "Properties": Object {
        "Description": "",
        "ManagedPolicyName": "ChangeOwnPasswordPolicy",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iam:ChangePassword",
                "iam:CreateAccessKey",
                "iam:CreateVirtualMFADevice",
                "iam:DeactivateMFADevice",
                "iam:DeleteAccessKey",
                "iam:DeleteVirtualMFADevice",
                "iam:EnableMFADevice",
                "iam:GetAccountPasswordPolicy",
                "iam:UpdateAccessKey",
                "iam:UpdateSigningCertificate",
                "iam:UploadSigningCertificate",
                "iam:UpdateLoginProfile",
                "iam:ResyncMFADevice",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":user/\${aws:username}",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":mfa/\${aws:username}",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "iam:ListUsers",
                "iam:ListVirtualMFADevices",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "DenyNetworkAccessPolicy11888F51": Object {
      "Properties": Object {
        "Description": "",
        "ManagedPolicyName": "DenyNetworkAccessPolicy",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2:AcceptVpcEndpointConnections",
                "ec2:AssociateDhcpOptions",
                "ec2:AssociateRouteTable",
                "ec2:AssociateSubnetCidrBlock",
                "ec2:AssociateVpcCidrBlock",
                "ec2:AttachInternetGateway",
                "ec2:AttachNetworkInterface",
                "ec2:AttachVpnGateway",
                "ec2:CreateCustomerGateway",
                "ec2:CreateDefaultSubnet",
                "ec2:CreateDefaultVpc",
                "ec2:CreateDhcpOptions",
                "ec2:CreateEgressOnlyInternetGateway",
                "ec2:CreateInternetGateway",
                "ec2:CreateNetworkAcl",
                "ec2:CreateNetworkAclEntry",
                "ec2:CreateRoute",
                "ec2:CreateRouteTable",
                "ec2:CreateSecurityGroup",
                "ec2:CreateSubnet",
                "ec2:CreateVpc",
                "ec2:CreateVpcEndpoint",
                "ec2:CreateVpcEndpointConnectionNotification",
                "ec2:CreateVpcEndpointServiceConfiguration",
                "ec2:CreateVpnConnection",
                "ec2:CreateVpnConnectionRoute",
                "ec2:CreateVpnGateway",
                "ec2:DeleteEgressOnlyInternetGateway",
                "ec2:DeleteNatGateway",
                "ec2:DeleteNetworkInterface",
                "ec2:DeleteNetworkInterfacePermission",
                "ec2:DeletePlacementGroup",
                "ec2:DeleteSubnet",
                "ec2:DeleteVpc",
                "ec2:DeleteVpcEndpointConnectionNotifications",
                "ec2:DeleteVpcEndpoints",
                "ec2:DeleteVpcEndpointServiceConfigurations",
                "ec2:DeleteVpnConnection",
                "ec2:DeleteVpnConnectionRoute",
                "ec2:DeleteVpnGateway",
                "ec2:DetachInternetGateway",
                "ec2:DetachNetworkInterface",
                "ec2:DetachVpnGateway",
                "ec2:DisableVgwRoutePropagation",
                "ec2:DisableVpcClassicLinkDnsSupport",
                "ec2:DisassociateRouteTable",
                "ec2:DisassociateSubnetCidrBlock",
                "ec2:DisassociateVpcCidrBlock",
                "ec2:EnableVgwRoutePropagation",
                "ec2:EnableVpcClassicLinkDnsSupport",
                "ec2:ModifyNetworkInterfaceAttribute",
                "ec2:ModifySubnetAttribute",
                "ec2:ModifyVpcAttribute",
                "ec2:ModifyVpcEndpoint",
                "ec2:ModifyVpcEndpointConnectionNotification",
                "ec2:ModifyVpcEndpointServiceConfiguration",
                "ec2:ModifyVpcEndpointServicePermissions",
                "ec2:ModifyVpcPeeringConnectionOptions",
                "ec2:ModifyVpcTenancy",
                "ec2:MoveAddressToVpc",
                "ec2:RejectVpcEndpointConnections",
                "ec2:ReplaceNetworkAclAssociation",
                "ec2:ReplaceNetworkAclEntry",
                "ec2:ReplaceRoute",
                "ec2:ReplaceRouteTableAssociation",
                "ec2:ResetNetworkInterfaceAttribute",
                "ec2:RestoreAddressToClassic",
                "ec2:UpdateSecurityGroupRuleDescriptionsEgress",
                "ec2:UpdateSecurityGroupRuleDescriptionsIngress",
                "directconnect:*",
              ],
              "Effect": "Deny",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "ec2:AcceptVpcPeeringConnection",
                "ec2:AttachClassicLinkVpc",
                "ec2:AuthorizeSecurityGroupEgress",
                "ec2:AuthorizeSecurityGroupIngress",
                "ec2:CreateVpcPeeringConnection",
                "ec2:DeleteCustomerGateway",
                "ec2:DeleteDhcpOptions",
                "ec2:DeleteInternetGateway",
                "ec2:DeleteNetworkAcl",
                "ec2:DeleteNetworkAclEntry",
                "ec2:DeleteRoute",
                "ec2:DeleteRouteTable",
                "ec2:DeleteSecurityGroup",
                "ec2:DeleteVolume",
                "ec2:DeleteVpcPeeringConnection",
                "ec2:DetachClassicLinkVpc",
                "ec2:DisableVpcClassicLink",
                "ec2:EnableVpcClassicLink",
                "ec2:GetConsoleScreenshot",
                "ec2:RejectVpcPeeringConnection",
                "ec2:RevokeSecurityGroupEgress",
                "ec2:RevokeSecurityGroupIngress",
                "ec2:AcceptTransitGatewayVpcAttachment",
                "ec2:AssociateTransitGatewayRouteTable",
                "ec2:CreateTransitGateway",
                "ec2:CreateTransitGatewayRoute",
                "ec2:CreateTransitGatewayRouteTable",
                "ec2:CreateTransitGatewayVpcAttachment",
                "ec2:DeleteTransitGateway",
                "ec2:DeleteTransitGatewayRoute",
                "ec2:DeleteTransitGatewayRouteTable",
                "ec2:DeleteTransitGatewayVpcAttachment",
                "ec2:DescribeTransitGatewayAttachments",
                "ec2:DescribeTransitGatewayRouteTables",
                "ec2:DescribeTransitGatewayVpcAttachments",
                "ec2:DescribeTransitGateways",
                "ec2:DisableTransitGatewayRouteTablePropagation",
                "ec2:DisassociateTransitGatewayRouteTable",
                "ec2:EnableTransitGatewayRouteTablePropagation",
                "ec2:ExportTransitGatewayRoutes",
                "ec2:GetTransitGatewayAttachmentPropagations",
                "ec2:GetTransitGatewayRouteTableAssociations",
                "ec2:GetTransitGatewayRouteTablePropagations",
                "ec2:ModifyTransitGatewayVpcAttachment",
                "ec2:RejectTransitGatewayVpcAttachment",
                "ec2:ReplaceTransitGatewayRoute",
                "ec2:SearchTransitGatewayRoutes",
              ],
              "Effect": "Deny",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "DeveloperGroup54FA9D31": Object {
      "Properties": Object {
        "GroupName": "Developers",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/PowerUserAccess",
              ],
            ],
          },
          Object {
            "Ref": "IpRestrictionAndMFAForcePolicyA06EC28F",
          },
          Object {
            "Ref": "ChangeOwnPasswordPolicy67755262",
          },
          Object {
            "Ref": "DenyNetworkAccessPolicy11888F51",
          },
        ],
      },
      "Type": "AWS::IAM::Group",
    },
    "IpRestrictionAndMFAForcePolicyA06EC28F": Object {
      "Properties": Object {
        "Description": "",
        "ManagedPolicyName": "IpRestrictionAndMFAForcePolicy",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "NotIpAddress": Object {
                  "aws:SourceIp": Array [
                    "8.8.8.8/32",
                  ],
                },
              },
              "Effect": "Deny",
              "Resource": "*",
            },
            Object {
              "Condition": Object {
                "BoolIfExists": Object {
                  "aws:MultiFactorAuthPresent": "false",
                },
              },
              "Effect": "Deny",
              "NotAction": "iam:*",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "iam:ListUsers",
                "iam:ListVirtualMFADevices",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "NetworkAdminGroup9CA8A5AE": Object {
      "Properties": Object {
        "GroupName": "NetworkAdmins",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/job-function/NetworkAdministrator",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/ReadOnlyAccess",
              ],
            ],
          },
          Object {
            "Ref": "IpRestrictionAndMFAForcePolicyA06EC28F",
          },
          Object {
            "Ref": "ChangeOwnPasswordPolicy67755262",
          },
        ],
      },
      "Type": "AWS::IAM::Group",
    },
    "OperatorGroup0CECA8D9": Object {
      "Properties": Object {
        "GroupName": "Operators",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AWSSupportAccess",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/ReadOnlyAccess",
              ],
            ],
          },
          Object {
            "Ref": "IpRestrictionAndMFAForcePolicyA06EC28F",
          },
          Object {
            "Ref": "ChangeOwnPasswordPolicy67755262",
          },
        ],
      },
      "Type": "AWS::IAM::Group",
    },
    "user2C2B57AE": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "AdminGroupCD52047A",
          },
        ],
        "UserName": "hoge",
      },
      "Type": "AWS::IAM::User",
    },
  },
}
`;
